### PR TITLE
Add REST Api for Claim Management

### PR DESCRIPTION
**Pull request by janakamarasena**
## Purpose
Resolves https://github.com/wso2/product-is/issues/5967 

Commit with non-generated code https://github.com/wso2/identity-api-server/pull/16/commits/697f57401d4dfc124d911feecf5af72667c4cd23

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.api.server.claim.management.common/src/main/java/org/wso2/carbon/identity/api/server/claim/management/common/Constant.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.ResourceBundle;
+
+/**
+ * Claim Management constant class.
+ */
+public class Constant {
+
+    public static final String CHALLENGE_QUESTION_PREFIX = "CMT-";
```
Comment (pamodaaw) [inline components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.api.server.claim.management.common/src/main/java/org/wso2/carbon/identity/api/server/claim/management/common/Constant.java]: Shall we change this constant name relevant to the claim management component

File: components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.api.server.claim.management.common/src/main/java/org/wso2/carbon/identity/api/server/claim/management/common/Constant.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+                "External claim URI cannot be empty."),
+        ERROR_CODE_INVALID_EXTERNAL_CLAIM_DIALECT("500031",
+                "Invalid external claim dialect.",
+                "The provided claim dialect is the local claim dialect and cannot be used as an external dialect."),
+        ERROR_CODE_EMPTY_EXTERNAL_DIALECT_URI("500032",
+                "Empty external claim dialect URI.",
+                "External dialect URI cannot be empty."),
+        ERROR_CODE_MAPPED_TO_EMPTY_LOCAL_CLAIM_URI("500033",
+                "Empty mapped local claim URI.",
+                "Mapped local claim URI cannot be empty"),
```
Comment (pamodaaw) [inline components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.api.server.claim.management.common/src/main/java/org/wso2/carbon/identity/api/server/claim/management/common/Constant.java]: Please add '.' at the end

File: components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/gen/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/ClaimManagementApiService.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import java.util.List;
+
+import java.io.InputStream;
+import org.apache.cxf.jaxrs.ext.multipart.Attachment;
+
+import javax.ws.rs.core.Response;
+
+public abstract class ClaimManagementApiService {
+    public abstract Response addClaimDialect(ClaimDialectReqDTO claimDialect);
+    public abstract Response addExternalClaim(String dialectId,ExternalClaimReqDTO externalClaim);
```
Comment (pamodaaw) [inline components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/gen/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/ClaimManagementApiService.java]: Missing spaces between arguments. Please check the other methods as well.

File: components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.api.server.claim.management.common/src/main/java/org/wso2/carbon/identity/api/server/claim/management/common/Constant.java (hunk 1 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.api.server.claim.management.common/src/main/java/org/wso2/carbon/identity/api/server/claim/management/common/Constant.java]: Fixed.

File: components/org.wso2.carbon.identity.api.server.claim.management/pom.xml [hunk 4]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (omindu) [inline components/org.wso2.carbon.identity.api.server.claim.management/pom.xml]: Missing licence header

File: components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.api.server.claim.management.common/src/main/java/org/wso2/carbon/identity/api/server/claim/management/common/Constant.java (hunk 2 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.api.server.claim.management.common/src/main/java/org/wso2/carbon/identity/api/server/claim/management/common/Constant.java]: Fixed.

File: components/org.wso2.carbon.identity.api.server.claim.management/pom.xml (hunk 4 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.api.server.claim.management/pom.xml]: Fixed.

File: components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java [hunk 5]
```diff
@@ -0,0 +1,796 @@
...
+    /**
+     * Add a local claim.
+     *
+     * @param localClaimReqDTO localClaimReqDTO.
+     * @return Resource identifier.
+     */
+    public String addLocalClaim(LocalClaimReqDTO localClaimReqDTO) {
+
+        try {
+
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java]: remove line

File: components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java [hunk 6]
```diff
@@ -0,0 +1,796 @@
...
+     * the resource id of the claim dialect changes when an update occurs.
+     *
+     * @param dialectId          dialectId.
+     * @param claimDialectReqDTO claimDialectReqDTO.
+     * @return Resource identifier.
+     */
+    public String updateClaimDialect(String dialectId, ClaimDialectReqDTO claimDialectReqDTO) {
+
+        try {
+
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java]: unnecessary line

File: components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java [hunk 7]
```diff
@@ -0,0 +1,796 @@
...
+            List<ClaimDialect> claimDialectList = getClaimMetadataManagementService().getClaimDialects(
+                    ContextLoader.getTenantDomainFromContext());
+            ClaimDialect claimDialect = extractDialectFromDialectList(base64DecodeId(dialectId), claimDialectList);
+
+            if (claimDialect == null) {
+                throw handleClaimManagementClientError(ERROR_CODE_DIALECT_NOT_FOUND, NOT_FOUND);
+            }
+
+            return getClaimDialectResDTO(claimDialect);
+
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java]: unnecessary line

File: components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java [hunk 8]
```diff
@@ -0,0 +1,796 @@
...
+        } catch (ClaimMetadataException e) {
+            throw handleClaimManagementException(e, ERROR_CODE_ERROR_RETRIEVING_DIALECT);
+        }
+
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java]: unnecessary line

File: components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java [hunk 9]
```diff
@@ -0,0 +1,796 @@
...
+        try {
+            List<ClaimDialect> claimDialectList = getClaimMetadataManagementService().getClaimDialects(
+                    ContextLoader.getTenantDomainFromContext());
+
+            return getClaimDialectResDTOs(claimDialectList);
+
+        } catch (ClaimMetadataException e) {
+            throw handleClaimManagementException(e, ERROR_CODE_ERROR_RETRIEVING_DIALECTS);
+        }
+
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java]: unnecessary line

File: components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java [hunk 10]
```diff
@@ -0,0 +1,796 @@
...
+        return getResourceId(localClaimReqDTO.getClaimURI());
+    }
+
+    /**
+     * Delete a local claim.
+     *
+     * @param claimId claimId.
+     * @return True if delete was successful without any errors.
+     */
+    public boolean deleteLocalClaim(String claimId) {
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java]: do we need boolean here?

File: components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java [hunk 11]
```diff
@@ -0,0 +1,796 @@
...
+            }
+
+            ExternalClaim externalClaim = extractExternalClaimFromClaimList(base64DecodeId(claimId), externalClaimList);
+
+            if (externalClaim == null) {
+                throw handleClaimManagementClientError(ERROR_CODE_EXTERNAL_CLAIM_NOT_FOUND, NOT_FOUND);
+            }
+
+            return getExternalClaimResDTO(externalClaim);
+
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java]: unnecessary line

File: components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java [hunk 12]
```diff
@@ -0,0 +1,796 @@
...
+        try {
+            List<ExternalClaim> externalClaimList = getClaimMetadataManagementService().getExternalClaims(
+                    base64DecodeId(dialectId),
+                    ContextLoader.getTenantDomainFromContext());
+            if (CollectionUtils.isEmpty(externalClaimList)) {
+                throw handleClaimManagementClientError(ERROR_CODE_CLAIMS_NOT_FOUND_FOR_DIALECT, NOT_FOUND);
+            }
+
+            return getExternalClaimResDTOs(externalClaimList);
+
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java]: unnecessary line

File: components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java (hunk 10 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java]: This is not necessary. Added to follow the pattern we have used when implementing the other server apis.

File: components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java (hunk 12 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java]: Fixed.

File: components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java (hunk 11 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java]: Fixed.

File: components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java (hunk 9 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java]: Fixed.

File: components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java (hunk 8 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java]: Fixed.

File: components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java (hunk 7 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java]: Fixed.

File: components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java (hunk 6 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java]: Fixed.

File: components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java (hunk 5 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.api.server.claim.management/org.wso2.carbon.identity.rest.api.server.claim.management.v1/src/main/java/org/wso2/carbon/identity/rest/api/server/claim/management/v1/core/ServerClaimManagementService.java]: Fixed.

**Pull request merged**